File: Advanced/PageAndRDL/CustomResourceLocator/VB.NET/CustomResourceLocator/MyPicturesLocator.vb
Imports GrapeCity.ActiveReports.Extensibility
Imports System.Globalization
Imports System.IO
Imports System.Runtime.InteropServices

Public Class MyPicturesLocator
	Inherits ResourceLocator

	Const UriSchemeMyImages As String = "MyPictures:"

	'<summary>
	'Look for the resources in My Pictures folder. 
	'</summary>
	'<param name="resourceInfo">The information about the resource to be obtained. </param>
	'<returns>The resource, null if it was not found. </returns>
	Public Overrides Function GetResource(ByVal resourceInfo As ResourceInfo) As Resource
		Dim resource As Resource
		Dim name As String = resourceInfo.Name

		If (String.IsNullOrEmpty(name)) Then

			Throw New ArgumentException(My.Resources.ResourceNameIsNull, "name")

		End If

		Dim uri As New Uri(name)
		If (uri.GetLeftPart(UriPartial.Scheme).StartsWith(UriSchemeMyImages, True, CultureInfo.InvariantCulture)) Then

			Dim stream As Stream = GetPictureFromSpecialFolder(uri)
			If (stream Is Nothing) Then
				stream = New MemoryStream()
				My.Resources.NoImage.Save(stream, My.Resources.NoImage.RawFormat)
			End If
			resource = New Resource(stream, uri)

		Else
			Throw New InvalidOperationException(My.Resources.ResourceSchemeIsNotSupported)
		End If
		Return resource
	End Function

	Function GetPictureFromSpecialFolder(ByVal path As Uri) As Stream
		Dim startPathPos As Integer = UriSchemeMyImages.Length

		If (startPathPos >= path.ToString().Length) Then
			Return Nothing
		End If

		Dim pictureName As String = path.ToString().Substring(startPathPos)
		Dim myPicturesPath As String = Environment.GetFolderPath(Environment.SpecialFolder.MyPictures)

		If (Not myPicturesPath.EndsWith("\\")) Then
			myPicturesPath += "\\"
		End If

		Dim picturePath As String = System.IO.Path.Combine(myPicturesPath, pictureName)

		If (Not File.Exists(picturePath)) Then
			Return Nothing
		End If

		Dim stream As New MemoryStream()

		Try
			Dim picture As Image = Image.FromFile(picturePath)
			picture.Save(stream, picture.RawFormat)
			stream.Position = 0
		Catch ex As OutOfMemoryException ' The file is not valid image, or GDI+ doesn't support such images.
			Return Nothing
		Catch ex As ExternalException
			Return Nothing
		End Try

		Return stream
	End Function

End Class

File: Advanced/PageAndRDL/CustomResourceLocator/C#/CustomResourceLocator/MyPicturesLocator.cs
using System;
using System.Drawing;
using System.Globalization;
using System.IO;
using System.Runtime.InteropServices;
using GrapeCity.ActiveReports.Extensibility;
using GrapeCity.ActiveReports.Samples.CustomResourceLocator.Properties;

namespace GrapeCity.ActiveReports.Samples.CustomResourceLocator
{
	/// <summary>
	/// Look for the resources in My Pictures folder. <see cref="ResourceLocator"/> 
	/// </summary>
	internal sealed class MyPicturesLocator : ResourceLocator
	{
		private const string UriSchemeMyImages = "MyPictures:";
		/// <summary>
		/// Obtain and return the resource. 
		/// </summary>
		/// <param name="resourceInfo">The information about the resource to be obtained. </param>
		/// <returns>The resource, null if it was not found. </returns>
		public override Resource GetResource(ResourceInfo resourceInfo)
		{
			Resource resource;
			string name = resourceInfo.Name;
			if (string.IsNullOrEmpty(name))
			{
				throw new ArgumentException(Resources.ResourceNameIsNull, "resourceInfo");
			}
			Uri uri = new Uri(name);
			if (uri.GetLeftPart(UriPartial.Scheme).StartsWith(UriSchemeMyImages, true, CultureInfo.InvariantCulture))
			{
				Stream stream = GetPictureFromSpecialFolder(uri);
				if (stream == null)
				{
					stream = new MemoryStream();
					Resources.NoImage.Save(stream, Resources.NoImage.RawFormat);
				}
				resource = new Resource(stream, uri);
			}
			else
			{
				throw new InvalidOperationException(Resources.ResourceSchemeIsNotSupported);
			}
			return resource;
		}

		/// <summary>
		/// Returns the specified image from My Pictures folder. 
		/// </summary>
		/// <param name="path">The uri of the image located in My Pictures code, i.e. MyImages:logo.gif.</param>
		/// <returns>The stream contains the image data or null if the picture can't be found or handled.</returns>
		private static Stream GetPictureFromSpecialFolder(Uri path)
		{
			int startPathPos = UriSchemeMyImages.Length;
			if (startPathPos >= path.ToString().Length)
			{
				return null;
			}
			string pictureName = path.ToString().Substring(startPathPos);
			string myPicturesPath = Environment.GetFolderPath(Environment.SpecialFolder.MyPictures);
			if (!myPicturesPath.EndsWith("\\")) myPicturesPath += "\\";
			 string picturePath = Path.Combine(myPicturesPath, pictureName);
			if (!File.Exists(picturePath)) return null;
			MemoryStream stream = new MemoryStream();
			try
			{
				Image picture = Image.FromFile(picturePath);
				picture.Save(stream, picture.RawFormat);
				stream.Position = 0;
			}
			catch (OutOfMemoryException) // The file is not valid image, or GDI+ doesn't support such images.
			{
				return null;				
			}
			catch (ExternalException)
			{
				return null;
			}
			return stream;
		}
	}
}
